_posts/2017-10-11-cross-site-request-forgery.markdown

```diff
@@ -0,0 +1,217 @@
+
+So all in all, when it comes to REST APIs, implementing CSRF protections is not really that straightforward and requires careful design process. Combining one of the above approaches with Origin/Referer header validation can provide a decent level of protection against CSRF attacks at REST APIs, however I've got to say that sounds like a lot of work...
+
+### Future of CSRF attacks
+
+Fortunately CSRF doesn't have a long term future, the currently draft spec of [SameSite] cookies will prevent these attacks completely. The new SameSite flag on cookies instructs the browser to only include the cookie values when the top level navigation origin matches the cookie's domain. The support for SameSite cookies is [work in progress][caniuse], but the future looks bright at least.
+
+### Takeaway
+
+Cross Site Request Forgery attacks are quite tricky to deal with, but there are several well established methods to prevent them fortunately. In my opinion the following best practices can help one secure their applications:
+
+* To limit content injection, disable [Content Type Sniffing] in your application.
+* Use either X-Frame-Options of Content-Security-Policy headers with frame-ancestors.
+* To limit Cross Site Scripting attacks that could be used for content injection or stealing CSRF tokens, set [X-XSS-Protection] header.
+* Use HttpOnly and Secure cookies wherever you can, they really make your applications more secure.
+* Review your CORS setup and try to only set the headers for the resources that you actually want to make available cross domain. Try to separate your web frontend from backend APIs to allow more fine-tuned CORS configurations.
+* Use CSRF tokens in your forms and consider implementing Origin/Referer header validation for your REST endpoints.
+
+Most of the above are actually very easy to do (see also [securityheaders.io]), but to really protect against CSRF attacks (especially when implementing REST APIs) one really has to consider the different use-cases for their application.
+
+[HTTP methods]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods
+[Content Injection]: https://www.owasp.org/index.php/Content_Spoofing
+[X-Frame-Options]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options
+[Content-Security-Policy]: https://content-security-policy.com/
+[CSP frame ancestors]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors
+[Referrer policy]: https://www.w3.org/TR/referrer-policy/
+[OWASP]: https://www.owasp.org/index.php/Cross-Site_Request_Forgery_%28CSRF%29_Prevention_Cheat_Sheet
+[Path Relative StyleSheet Import]: http://blog.portswigger.net/2015/02/prssi.html
```

Comment:
Wow, this is a rabbit hole of terrifying CSS data exfiltration techniques.

Reply:
It is a really interesting issue indeed :)